bindings/rust/s2n-tls/src/renegotiate.rs

```diff
@@ -87,16 +294,77 @@ impl Connection {
     ///
     /// See s2n_renegotiate in [the C API documentation](https://github.com/aws/s2n-tls/blob/main/api/unstable/renegotiate.h).
     pub fn poll_renegotiate(&mut self, buf: &mut [u8]) -> (Poll<Result<(), Error>>, usize) {
-        let mut blocked = s2n_blocked_status::NOT_BLOCKED;
         let buf_len: isize = buf.len().try_into().unwrap_or(0);
-        let buf_ptr = buf.as_mut_ptr();
-        let mut read: isize = 0;
+        let buf_ptr = buf.as_ptr() as *mut ::libc::c_void;
+        self.poll_renegotiate_raw(buf_ptr, buf_len)
+    }
 
-        let r = self.poll_negotiate_method(|conn| {
-            unsafe { s2n_renegotiate(conn.as_ptr(), buf_ptr, buf_len, &mut read, &mut blocked) }
-                .into_poll()
-        });
-        (r, read.try_into().unwrap())
+    /// Encrypts and sends data on a connection where
+    /// [negotiate](`Self::poll_negotiate`) has succeeded.
+    ///
+    /// Returns the number of bytes written, and may indicate a partial write.
+    pub fn poll_send(&mut self, buf: &[u8]) -> Poll<Result<usize, Error>> {
+        if self.is_renegotiating() {
+            return Ready(Err(Error::from(
+                ErrorType::Blocked,
+                "RenegotiateError",
+                "Cannot send application data while renegotiating",
+            )));
+        }
+        let mut blocked = s2n_blocked_status::NOT_BLOCKED;
+        let buf_len: isize = buf.len().try_into().map_err(|_| Error::INVALID_INPUT)?;
+        let buf_ptr = buf.as_ptr() as *const libc::c_void;
+        let result = unsafe { s2n_send(self.as_ptr(), buf_ptr, buf_len, &mut blocked) }.into_poll();
+        self.renegotiate_state_mut().send_pending = result.is_pending();
+        result
+    }
+
+    pub(crate) fn poll_recv_raw(
+        &mut self,
+        buf_ptr: *mut libc::c_void,
+        buf_len: isize,
+    ) -> Poll<Result<usize, Error>> {
+        if self.is_renegotiating() && self.peek_len() > 0 {
```

Comment:
Tricky. 

Technically yes, if we [fix poll_flush](https://github.com/aws/s2n-tls/issues/4803). But s2n_renegotiate_wipe checking the IO buffers was always just intended as a best-effort check for pending writes, which also involve other connection state, like [conn->current_user_data_consumed](https://github.com/aws/s2n-tls/blob/main/tls/s2n_send.c#L188). The idea was that we didn't want to try to preserve all the state we'd need to continue re-trying a send after the wipe. Since we just wipe that state, the behavior of poll_send after the successful wipe would be undefined and maybe incorrect. If we fix poll_flush, I might need to make the s2n_renegotiate_wipe checks in the C library stricter.